parts/linux/cloud-init/artifacts/localdns.sh

```diff
@@ -0,0 +1,496 @@
+    if [ ! -z "${COREDNS_PID:-}" ]; then
+        # Check if the process exists by using `ps`.
+        if ps -p "${COREDNS_PID}" >/dev/null 2>&1; then
+            if [[ "${LOCALDNS_SHUTDOWN_DELAY}" -gt 0 ]]; then
+                # Wait after removing iptables rules and DNS configuration so that we can let connections transition.
+                echo "Sleeping ${LOCALDNS_SHUTDOWN_DELAY} seconds to allow connections to terminate."
+                sleep "${LOCALDNS_SHUTDOWN_DELAY}"
+            fi
+            echo "Sending SIGINT to localdns and waiting for it to terminate."
+
+            # Send SIGINT to localdns to trigger a graceful shutdown.
+            kill -SIGINT "${COREDNS_PID}"
+            kill_status=$?
+            if [ $kill_status -eq 0 ]; then
+                echo "Successfully sent SIGINT to localdns."
+            else
+                echo "Failed to send SIGINT to localdns. Exit status: $kill_status."
+                return 1
+            fi
+
+            # Wait for the process to terminate.
+            if wait "${COREDNS_PID}"; then
+                echo "Localdns terminated successfully."
+            else
+                echo "Localdns failed to terminate properly."
+                return 1
+            fi
+        fi
+    fi
+
+    # Delete the dummy interface if present.
+    if ip link show dev localdns >/dev/null 2>&1; then
+        echo "Removing localdns dummy interface."
+        ip link del name localdns
+        if [ $? -eq 0 ]; then
+            echo "Successfully removed localdns dummy interface."
+        else
+            echo "Failed to remove localdns dummy interface."
+            return 1
+        fi
+    fi
+
+    # Indicate successful cleanup.
+    echo "Successfully cleanup localdns related configurations."
+    return 0
+}
+
+# Start the localdns watchdog.
+# This function is used to check the health of localdns and restart it if necessary.
+# It uses systemd-notify to send a watchdog ping to the systemd service manager.
+# The watchdog interval is set in the systemd unit file and is passed to the script via the WATCHDOG_USEC environment variable.
+# The health check is a DNS request to the localdns service IPs.
+# The health check is run at 20% of the WATCHDOG_USEC interval.
+# If the health check fails, the script will exit and systemd will restart the service.
+start_localdns_watchdog() {
+    if [[ -n "${NOTIFY_SOCKET:-}" && -n "${WATCHDOG_USEC:-}" ]]; then
+        # Health check at 20% of WATCHDOG_USEC; this means that we should check.
+        # five times in every watchdog interval, and thus need to fail five checks to get restarted.
+        HEALTH_CHECK_INTERVAL=$((${WATCHDOG_USEC:-5000000} * 20 / 100 / 1000000))
+        echo "Starting watchdog loop at ${HEALTH_CHECK_INTERVAL} second intervals."
+        while true; do
+            if [[ "$($CURL_COMMAND)" == "OK" ]]; then
+                if dig +short +timeout=1 +tries=1 -f <(printf "%s" "$HEALTH_CHECK_DNS_REQUEST"); then
```

Comment:
I think we can merge this if condition with the previous line? because you don't have an else case here.